pkg/util/clusterdata/clusterdata.go

```diff
@@ -0,0 +1,124 @@
+)
+
+// OpenShiftClusterEnricher must update the cluster object with
+// data received from API server
+type OpenShiftClusterEnricher interface {
+	Enrich(ctx context.Context, ocs ...*api.OpenShiftCluster)
+}
+
+type enricherTaskConstructor func(*logrus.Entry, *rest.Config, *api.OpenShiftCluster) (enricherTask, error)
+type enricherTask interface {
+	SetDefaults()
+	FetchData(chan<- func(), chan<- error)
+}
+
+// NewBestEffortEnricher returns an enricher that attempts to populate
+// fields, but ignores errors in case of failures
+func NewBestEffortEnricher(log *logrus.Entry, env env.Interface) OpenShiftClusterEnricher {
+	return &bestEffortEnricher{
+		log: log,
+		env: env,
+
+		restConfig: restconfig.RestConfig,
+		taskConstructors: []enricherTaskConstructor{
+			newClusterVersionEnricherTask,
+			newServicePrincipalEnricherTask,
+			newWorkerProfilesEnricherTask,
+		},
+	}
+}
+
+type bestEffortEnricher struct {
+	log *logrus.Entry
+	env env.Interface
+
+	restConfig       func(env env.Interface, oc *api.OpenShiftCluster) (*rest.Config, error)
+	taskConstructors []enricherTaskConstructor
+}
+
+func (e *bestEffortEnricher) Enrich(ctx context.Context, ocs ...*api.OpenShiftCluster) {
+	var wg sync.WaitGroup
+	wg.Add(len(ocs))
+	for i := range ocs {
+		go func(i int) {
+			defer wg.Done()
+			e.enrichOne(ctx, ocs[i])
+		}(i) // https://golang.org/doc/faq#closures_and_goroutines
+	}
+	wg.Wait()
+}
+
+func (e *bestEffortEnricher) enrichOne(ctx context.Context, oc *api.OpenShiftCluster) {
+	restConfig, err := e.restConfig(e.env, oc)
+	if err != nil {
+		e.log.Error(err)
+		return
+	}
+
+	// TODO: Get rid of the wrapping RoundTripper once implementation of the KEP below lands into openshift/client-go:
+	//       https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20200123-client-go-ctx.md
+	restConfig.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (*http.Response, error) {
+			return rt.RoundTrip(req.WithContext(ctx))
+		})
+	})
+
+	tasks := make([]enricherTask, 0, len(e.taskConstructors))
+	for i := range e.taskConstructors {
+		task, err := e.taskConstructors[i](e.log, restConfig, oc)
+		if err != nil {
+			e.log.Error(err)
+			continue
+		}
+		tasks = append(tasks, task)
+		task.SetDefaults()
+	}
+
+	// We must use a buffered channels with length equal to the number of tasks
+	// to ensure that FetchData goroutines return as soon as they finish
+	// their job and do not wait for someone to read from the channel.
+	// Otherwise in case of error in one of the goroutines or on timeout
+	// they will not be garbage collected.
+	callbacks := make(chan func(), len(tasks))
+	errors := make(chan error, len(tasks))
+	for i := range tasks {
+		go tasks[i].FetchData(callbacks, errors)
+	}
+
+out:
+	for i := 0; i < len(tasks); i++ {
+		select {
+		case f := <-callbacks:
+			f()
+		case <-errors:
+			// Ignore errors. We log them in each task locally
```

Comment:
@olga-mirensky this is an interesting case as you investigate using github.com/pkg/errors.  By including stack traces in our errors we could remove the error logging in the individual tasks and log centrally here.